// llamaindex/dist/storage/vectorStore/types.js
import { getEmbeddedModel } from "../../internal/settings/EmbedModel.js";
export var VectorStoreQueryMode;
(function(VectorStoreQueryMode) {
    VectorStoreQueryMode["DEFAULT"] = "default";
    VectorStoreQueryMode["SPARSE"] = "sparse";
    VectorStoreQueryMode["HYBRID"] = "hybrid";
    // fit learners
    VectorStoreQueryMode["SVM"] = "svm";
    VectorStoreQueryMode["LOGISTIC_REGRESSION"] = "logistic_regression";
    VectorStoreQueryMode["LINEAR_REGRESSION"] = "linear_regression";
    // maximum marginal relevance
    VectorStoreQueryMode["MMR"] = "mmr";
})(VectorStoreQueryMode || (VectorStoreQueryMode = {}));
export var FilterOperator;
(function(FilterOperator) {
    FilterOperator["EQ"] = "==";
    FilterOperator["IN"] = "in";
    FilterOperator["GT"] = ">";
    FilterOperator["LT"] = "<";
    FilterOperator["NE"] = "!=";
    FilterOperator["GTE"] = ">=";
    FilterOperator["LTE"] = "<=";
    FilterOperator["NIN"] = "nin";
    FilterOperator["ANY"] = "any";
    FilterOperator["ALL"] = "all";
    FilterOperator["TEXT_MATCH"] = "text_match";
    FilterOperator["CONTAINS"] = "contains";
})(FilterOperator || (FilterOperator = {}));
export var FilterCondition;
(function(FilterCondition) {
    FilterCondition["AND"] = "and";
    FilterCondition["OR"] = "or";
})(FilterCondition || (FilterCondition = {}));
export class VectorStoreBase {
    embedModel;
    constructor(embedModel){
        this.embedModel = embedModel ?? getEmbeddedModel();
    }
}
