// llamaindex/dist/cloud/config.js
import { SentenceSplitter } from "@llamaindex/core/node-parser";
import { OpenAIEmbedding } from "../embeddings/OpenAIEmbedding.js";
function getTransformationConfig(transformation) {
    if (transformation instanceof SentenceSplitter) {
        return {
            configurable_transformation_type: "SENTENCE_AWARE_NODE_PARSER",
            component: {
                chunk_size: transformation.chunkSize,
                chunk_overlap: transformation.chunkOverlap,
                include_metadata: transformation.includeMetadata,
                include_prev_next_rel: transformation.includePrevNextRel
            }
        };
    }
    if (transformation instanceof OpenAIEmbedding) {
        return {
            configurable_transformation_type: "OPENAI_EMBEDDING",
            component: {
                model: transformation.model,
                api_key: transformation.apiKey,
                embed_batch_size: transformation.embedBatchSize,
                dimensions: transformation.dimensions
            }
        };
    }
    throw new Error(`Unsupported transformation: ${typeof transformation}`);
}
export async function getPipelineCreate(params) {
    const { pipelineName, pipelineType, transformations = [] } = params;
    return {
        name: pipelineName,
        configured_transformations: transformations.map(getTransformationConfig),
        pipeline_type: pipelineType
    };
}
