// llamaindex/dist/cjs/evaluation/Correctness.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "CorrectnessEvaluator", {
    enumerable: true,
    get: function() {
        return CorrectnessEvaluator;
    }
});
const _schema = require("@llamaindex/core/schema");
const _utils = require("@llamaindex/core/utils");
const _Mixin = require("../prompts/Mixin.js");
const _Settings = require("../Settings.js");
const _prompts = require("./prompts.js");
const _utils1 = require("./utils.js");
class CorrectnessEvaluator extends _Mixin.PromptMixin {
    scoreThreshold;
    parserFunction;
    llm;
    correctnessPrompt = _prompts.defaultCorrectnessSystemPrompt;
    constructor(params){
        super();
        this.llm = (0, _Settings.llmFromSettingsOrContext)(params?.serviceContext);
        this.correctnessPrompt = _prompts.defaultCorrectnessSystemPrompt;
        this.scoreThreshold = params?.scoreThreshold ?? 4.0;
        this.parserFunction = params?.parserFunction ?? _utils1.defaultEvaluationParser;
    }
    _updatePrompts(prompts) {
        if ("correctnessPrompt" in prompts) {
            this.correctnessPrompt = prompts["correctnessPrompt"];
        }
    }
    /**
   *
   * @param query Query to evaluate
   * @param response  Response to evaluate
   * @param contexts Array of contexts
   * @param reference  Reference response
   */ async evaluate({ query, response, contexts, reference }) {
        if (query === null || response === null) {
            throw new Error("query, and response must be provided");
        }
        const messages = [
            {
                role: "system",
                content: this.correctnessPrompt()
            },
            {
                role: "user",
                content: (0, _prompts.defaultUserPrompt)({
                    query: (0, _utils.extractText)(query),
                    generatedAnswer: response,
                    referenceAnswer: reference || "(NO REFERENCE ANSWER SUPPLIED)"
                })
            }
        ];
        const evalResponse = await this.llm.chat({
            messages
        });
        const [score, reasoning] = this.parserFunction((0, _utils.extractText)(evalResponse.message.content));
        return {
            query: query,
            response: response,
            passing: score >= this.scoreThreshold || score === null,
            score: score,
            feedback: reasoning
        };
    }
    /**
   * @param query Query to evaluate
   * @param response  Response to evaluate
   */ async evaluateResponse({ query, response }) {
        const responseStr = (0, _utils.extractText)(response?.message.content);
        const contexts = [];
        if (response) {
            for (const node of response.sourceNodes || []){
                contexts.push(node.node.getContent(_schema.MetadataMode.ALL));
            }
        }
        return this.evaluate({
            query,
            response: responseStr,
            contexts
        });
    }
}
