// llamaindex/dist/llm/huggingface.js
function applyDecs2203RFactory() {
    function createAddInitializerMethod(initializers, decoratorFinishedRef) {
        return function addInitializer(initializer) {
            assertNotFinished(decoratorFinishedRef, "addInitializer");
            assertCallable(initializer, "An initializer");
            initializers.push(initializer);
        };
    }
    function memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value) {
        var kindStr;
        switch(kind){
            case 1:
                kindStr = "accessor";
                break;
            case 2:
                kindStr = "method";
                break;
            case 3:
                kindStr = "getter";
                break;
            case 4:
                kindStr = "setter";
                break;
            default:
                kindStr = "field";
        }
        var ctx = {
            kind: kindStr,
            name: isPrivate ? "#" + name : name,
            static: isStatic,
            private: isPrivate,
            metadata: metadata
        };
        var decoratorFinishedRef = {
            v: false
        };
        ctx.addInitializer = createAddInitializerMethod(initializers, decoratorFinishedRef);
        var get, set;
        if (kind === 0) {
            if (isPrivate) {
                get = desc.get;
                set = desc.set;
            } else {
                get = function() {
                    return this[name];
                };
                set = function(v) {
                    this[name] = v;
                };
            }
        } else if (kind === 2) {
            get = function() {
                return desc.value;
            };
        } else {
            if (kind === 1 || kind === 3) {
                get = function() {
                    return desc.get.call(this);
                };
            }
            if (kind === 1 || kind === 4) {
                set = function(v) {
                    desc.set.call(this, v);
                };
            }
        }
        ctx.access = get && set ? {
            get: get,
            set: set
        } : get ? {
            get: get
        } : {
            set: set
        };
        try {
            return dec(value, ctx);
        } finally{
            decoratorFinishedRef.v = true;
        }
    }
    function assertNotFinished(decoratorFinishedRef, fnName) {
        if (decoratorFinishedRef.v) {
            throw new Error("attempted to call " + fnName + " after decoration was finished");
        }
    }
    function assertCallable(fn, hint) {
        if (typeof fn !== "function") {
            throw new TypeError(hint + " must be a function");
        }
    }
    function assertValidReturnValue(kind, value) {
        var type = typeof value;
        if (kind === 1) {
            if (type !== "object" || value === null) {
                throw new TypeError("accessor decorators must return an object with get, set, or init properties or void 0");
            }
            if (value.get !== undefined) {
                assertCallable(value.get, "accessor.get");
            }
            if (value.set !== undefined) {
                assertCallable(value.set, "accessor.set");
            }
            if (value.init !== undefined) {
                assertCallable(value.init, "accessor.init");
            }
        } else if (type !== "function") {
            var hint;
            if (kind === 0) {
                hint = "field";
            } else if (kind === 10) {
                hint = "class";
            } else {
                hint = "method";
            }
            throw new TypeError(hint + " decorators must return a function or void 0");
        }
    }
    function applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata) {
        var decs = decInfo[0];
        var desc, init, value;
        if (isPrivate) {
            if (kind === 0 || kind === 1) {
                desc = {
                    get: decInfo[3],
                    set: decInfo[4]
                };
            } else if (kind === 3) {
                desc = {
                    get: decInfo[3]
                };
            } else if (kind === 4) {
                desc = {
                    set: decInfo[3]
                };
            } else {
                desc = {
                    value: decInfo[3]
                };
            }
        } else if (kind !== 0) {
            desc = Object.getOwnPropertyDescriptor(base, name);
        }
        if (kind === 1) {
            value = {
                get: desc.get,
                set: desc.set
            };
        } else if (kind === 2) {
            value = desc.value;
        } else if (kind === 3) {
            value = desc.get;
        } else if (kind === 4) {
            value = desc.set;
        }
        var newValue, get, set;
        if (typeof decs === "function") {
            newValue = memberDec(decs, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
            if (newValue !== void 0) {
                assertValidReturnValue(kind, newValue);
                if (kind === 0) {
                    init = newValue;
                } else if (kind === 1) {
                    init = newValue.init;
                    get = newValue.get || value.get;
                    set = newValue.set || value.set;
                    value = {
                        get: get,
                        set: set
                    };
                } else {
                    value = newValue;
                }
            }
        } else {
            for(var i = decs.length - 1; i >= 0; i--){
                var dec = decs[i];
                newValue = memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
                if (newValue !== void 0) {
                    assertValidReturnValue(kind, newValue);
                    var newInit;
                    if (kind === 0) {
                        newInit = newValue;
                    } else if (kind === 1) {
                        newInit = newValue.init;
                        get = newValue.get || value.get;
                        set = newValue.set || value.set;
                        value = {
                            get: get,
                            set: set
                        };
                    } else {
                        value = newValue;
                    }
                    if (newInit !== void 0) {
                        if (init === void 0) {
                            init = newInit;
                        } else if (typeof init === "function") {
                            init = [
                                init,
                                newInit
                            ];
                        } else {
                            init.push(newInit);
                        }
                    }
                }
            }
        }
        if (kind === 0 || kind === 1) {
            if (init === void 0) {
                init = function(instance, init) {
                    return init;
                };
            } else if (typeof init !== "function") {
                var ownInitializers = init;
                init = function(instance, init) {
                    var value = init;
                    for(var i = 0; i < ownInitializers.length; i++){
                        value = ownInitializers[i].call(instance, value);
                    }
                    return value;
                };
            } else {
                var originalInitializer = init;
                init = function(instance, init) {
                    return originalInitializer.call(instance, init);
                };
            }
            ret.push(init);
        }
        if (kind !== 0) {
            if (kind === 1) {
                desc.get = value.get;
                desc.set = value.set;
            } else if (kind === 2) {
                desc.value = value;
            } else if (kind === 3) {
                desc.get = value;
            } else if (kind === 4) {
                desc.set = value;
            }
            if (isPrivate) {
                if (kind === 1) {
                    ret.push(function(instance, args) {
                        return value.get.call(instance, args);
                    });
                    ret.push(function(instance, args) {
                        return value.set.call(instance, args);
                    });
                } else if (kind === 2) {
                    ret.push(value);
                } else {
                    ret.push(function(instance, args) {
                        return value.call(instance, args);
                    });
                }
            } else {
                Object.defineProperty(base, name, desc);
            }
        }
    }
    function applyMemberDecs(Class, decInfos, metadata) {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var existingProtoNonFields = new Map();
        var existingStaticNonFields = new Map();
        for(var i = 0; i < decInfos.length; i++){
            var decInfo = decInfos[i];
            if (!Array.isArray(decInfo)) continue;
            var kind = decInfo[1];
            var name = decInfo[2];
            var isPrivate = decInfo.length > 3;
            var isStatic = kind >= 5;
            var base;
            var initializers;
            if (isStatic) {
                base = Class;
                kind = kind - 5;
                staticInitializers = staticInitializers || [];
                initializers = staticInitializers;
            } else {
                base = Class.prototype;
                protoInitializers = protoInitializers || [];
                initializers = protoInitializers;
            }
            if (kind !== 0 && !isPrivate) {
                var existingNonFields = isStatic ? existingStaticNonFields : existingProtoNonFields;
                var existingKind = existingNonFields.get(name) || 0;
                if (existingKind === true || existingKind === 3 && kind !== 4 || existingKind === 4 && kind !== 3) {
                    throw new Error("Attempted to decorate a public method/accessor that has the same name as a previously decorated public method/accessor. This is not currently supported by the decorators plugin. Property name was: " + name);
                } else if (!existingKind && kind > 2) {
                    existingNonFields.set(name, kind);
                } else {
                    existingNonFields.set(name, true);
                }
            }
            applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata);
        }
        pushInitializers(ret, protoInitializers);
        pushInitializers(ret, staticInitializers);
        return ret;
    }
    function pushInitializers(ret, initializers) {
        if (initializers) {
            ret.push(function(instance) {
                for(var i = 0; i < initializers.length; i++){
                    initializers[i].call(instance);
                }
                return instance;
            });
        }
    }
    function applyClassDecs(targetClass, classDecs, metadata) {
        if (classDecs.length > 0) {
            var initializers = [];
            var newClass = targetClass;
            var name = targetClass.name;
            for(var i = classDecs.length - 1; i >= 0; i--){
                var decoratorFinishedRef = {
                    v: false
                };
                try {
                    var nextNewClass = classDecs[i](newClass, {
                        kind: "class",
                        name: name,
                        addInitializer: createAddInitializerMethod(initializers, decoratorFinishedRef),
                        metadata
                    });
                } finally{
                    decoratorFinishedRef.v = true;
                }
                if (nextNewClass !== undefined) {
                    assertValidReturnValue(10, nextNewClass);
                    newClass = nextNewClass;
                }
            }
            return [
                defineMetadata(newClass, metadata),
                function() {
                    for(var i = 0; i < initializers.length; i++){
                        initializers[i].call(newClass);
                    }
                }
            ];
        }
    }
    function defineMetadata(Class, metadata) {
        return Object.defineProperty(Class, Symbol.metadata || Symbol.for("Symbol.metadata"), {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    return function applyDecs2203R(targetClass, memberDecs, classDecs, parentClass) {
        if (parentClass !== void 0) {
            var parentMetadata = parentClass[Symbol.metadata || Symbol.for("Symbol.metadata")];
        }
        var metadata = Object.create(parentMetadata === void 0 ? null : parentMetadata);
        var e = applyMemberDecs(targetClass, memberDecs, metadata);
        if (!classDecs.length) defineMetadata(targetClass, metadata);
        return {
            e: e,
            get c () {
                return applyClassDecs(targetClass, classDecs, metadata);
            }
        };
    };
}
function _apply_decs_2203_r(targetClass, memberDecs, classDecs, parentClass) {
    return (_apply_decs_2203_r = applyDecs2203RFactory())(targetClass, memberDecs, classDecs, parentClass);
}
var _initProto, _initProto1;
import { HfInference } from "@huggingface/inference";
import "@llamaindex/core/llms";
import { BaseLLM } from "@llamaindex/core/llms";
import { streamConverter, wrapLLMEvent } from "@llamaindex/core/utils";
import { lazyLoadTransformers } from "../internal/deps/transformers.js";
const DEFAULT_PARAMS = {
    temperature: 0.1,
    topP: 1,
    maxTokens: undefined,
    contextWindow: 3900
};
/**
    Wrapper on the Hugging Face's Inference API.
    API Docs: https://huggingface.co/docs/huggingface.js/inference/README
    List of tasks with models: huggingface.co/api/tasks 
    
    Note that Conversational API is not yet supported by the Inference API.
    They recommend using the text generation API instead.
    See: https://github.com/huggingface/huggingface.js/issues/586#issuecomment-2024059308
 */ export class HuggingFaceInferenceAPI extends BaseLLM {
    static{
        ({ e: [_initProto] } = _apply_decs_2203_r(this, [
            [
                wrapLLMEvent,
                2,
                "chat"
            ]
        ], []));
    }
    model;
    temperature;
    topP;
    maxTokens;
    contextWindow;
    hf;
    constructor(init){
        super();
        _initProto(this);
        const { model, temperature, topP, maxTokens, contextWindow, accessToken, endpoint, ...hfInferenceOpts } = init;
        this.hf = new HfInference(accessToken, hfInferenceOpts);
        this.model = model;
        this.temperature = temperature ?? DEFAULT_PARAMS.temperature;
        this.topP = topP ?? DEFAULT_PARAMS.topP;
        this.maxTokens = maxTokens ?? DEFAULT_PARAMS.maxTokens;
        this.contextWindow = contextWindow ?? DEFAULT_PARAMS.contextWindow;
        if (endpoint) this.hf.endpoint(endpoint);
    }
    get metadata() {
        return {
            model: this.model,
            temperature: this.temperature,
            topP: this.topP,
            maxTokens: this.maxTokens,
            contextWindow: this.contextWindow,
            tokenizer: undefined
        };
    }
    async chat(params) {
        if (params.stream) return this.streamChat(params);
        return this.nonStreamChat(params);
    }
    messagesToPrompt(messages) {
        let prompt = "";
        for (const message of messages){
            if (message.role === "system") {
                prompt += `<|system|>\n${message.content}</s>\n`;
            } else if (message.role === "user") {
                prompt += `<|user|>\n${message.content}</s>\n`;
            } else if (message.role === "assistant") {
                prompt += `<|assistant|>\n${message.content}</s>\n`;
            }
        }
        // ensure we start with a system prompt, insert blank if needed
        if (!prompt.startsWith("<|system|>\n")) {
            prompt = "<|system|>\n</s>\n" + prompt;
        }
        // add final assistant prompt
        prompt = prompt + "<|assistant|>\n";
        return prompt;
    }
    async nonStreamChat(params) {
        const res = await this.hf.textGeneration({
            model: this.model,
            inputs: this.messagesToPrompt(params.messages),
            parameters: this.metadata
        });
        return {
            raw: res,
            message: {
                content: res.generated_text,
                role: "assistant"
            }
        };
    }
    async *streamChat(params) {
        const stream = this.hf.textGenerationStream({
            model: this.model,
            inputs: this.messagesToPrompt(params.messages),
            parameters: this.metadata
        });
        yield* streamConverter(stream, (chunk)=>({
                delta: chunk.token.text,
                raw: chunk
            }));
    }
}
const DEFAULT_HUGGINGFACE_MODEL = "stabilityai/stablelm-tuned-alpha-3b";
export class HuggingFaceLLM extends BaseLLM {
    static{
        ({ e: [_initProto1] } = _apply_decs_2203_r(this, [
            [
                wrapLLMEvent,
                2,
                "chat"
            ]
        ], []));
    }
    modelName;
    tokenizerName;
    temperature;
    topP;
    maxTokens;
    contextWindow;
    tokenizer = null;
    model = null;
    constructor(init){
        super();
        _initProto1(this);
        this.modelName = init?.modelName ?? DEFAULT_HUGGINGFACE_MODEL;
        this.tokenizerName = init?.tokenizerName ?? DEFAULT_HUGGINGFACE_MODEL;
        this.temperature = init?.temperature ?? DEFAULT_PARAMS.temperature;
        this.topP = init?.topP ?? DEFAULT_PARAMS.topP;
        this.maxTokens = init?.maxTokens ?? DEFAULT_PARAMS.maxTokens;
        this.contextWindow = init?.contextWindow ?? DEFAULT_PARAMS.contextWindow;
    }
    get metadata() {
        return {
            model: this.modelName,
            temperature: this.temperature,
            topP: this.topP,
            maxTokens: this.maxTokens,
            contextWindow: this.contextWindow,
            tokenizer: undefined
        };
    }
    async getTokenizer() {
        const { AutoTokenizer } = await lazyLoadTransformers();
        if (!this.tokenizer) {
            this.tokenizer = await AutoTokenizer.from_pretrained(this.tokenizerName);
        }
        return this.tokenizer;
    }
    async getModel() {
        const { AutoModelForCausalLM } = await lazyLoadTransformers();
        if (!this.model) {
            this.model = await AutoModelForCausalLM.from_pretrained(this.modelName);
        }
        return this.model;
    }
    async chat(params) {
        if (params.stream) return this.streamChat(params);
        return this.nonStreamChat(params);
    }
    async nonStreamChat(params) {
        const tokenizer = await this.getTokenizer();
        const model = await this.getModel();
        const messageInputs = params.messages.map((msg)=>({
                role: msg.role,
                content: msg.content
            }));
        const inputs = tokenizer.apply_chat_template(messageInputs, {
            add_generation_prompt: true,
            ...this.metadata
        });
        // TODO: the input for model.generate should be updated when using @xenova/transformers v3
        // We should add `stopping_criteria` also when it's supported in v3
        // See: https://github.com/xenova/transformers.js/blob/3260640b192b3e06a10a1f4dc004b1254fdf1b80/src/models.js#L1248C9-L1248C27
        const outputs = await model.generate(inputs, this.metadata);
        const outputText = tokenizer.batch_decode(outputs, {
            skip_special_tokens: false
        });
        return {
            raw: outputs,
            message: {
                content: outputText.join(""),
                role: "assistant"
            }
        };
    }
    async *streamChat(params) {
        // @xenova/transformers v2 doesn't support streaming generation yet
        // they are working on it in v3
        // See: https://github.com/xenova/transformers.js/blob/3260640b192b3e06a10a1f4dc004b1254fdf1b80/src/models.js#L1249
        throw new Error("Method not implemented.");
    }
}
