// llamaindex/dist/cjs/tools/WikipediaTool.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "WikipediaTool", {
    enumerable: true,
    get: function() {
        return WikipediaTool;
    }
});
const _wikipedia = /*#__PURE__*/ _interop_require_default(require("wikipedia"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const DEFAULT_META_DATA = {
    name: "wikipedia_tool",
    description: "A tool that uses a query engine to search Wikipedia.",
    parameters: {
        type: "object",
        properties: {
            query: {
                type: "string",
                description: "The query to search for"
            },
            lang: {
                type: "string",
                description: "The language to search in",
                nullable: true
            }
        },
        required: [
            "query"
        ]
    }
};
class WikipediaTool {
    DEFAULT_LANG = "en";
    metadata;
    constructor(params){
        this.metadata = params?.metadata || DEFAULT_META_DATA;
    }
    async loadData(page, lang = this.DEFAULT_LANG) {
        _wikipedia.default.default.setLang(lang);
        const pageResult = await _wikipedia.default.default.page(page, {
            autoSuggest: false
        });
        const content = await pageResult.content();
        return content;
    }
    async call({ query, lang = this.DEFAULT_LANG }) {
        const searchResult = await _wikipedia.default.default.search(query);
        if (searchResult.results.length === 0) return "No search results.";
        return await this.loadData(searchResult.results[0].title, lang);
    }
}
