// @llamaindex/env/dist/cjs/tokenizers/js.js
// Note: js-tiktoken it's 60x slower than the WASM implementation - use it only for unsupported environments
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Tokenizers: function() {
        return _types.Tokenizers;
    },
    tokenizers: function() {
        return tokenizers;
    }
});
const _jstiktoken = require("js-tiktoken");
const _types = require("./types.js");
class TokenizerSingleton {
    defaultTokenizer;
    constructor(){
        const encoding = (0, _jstiktoken.getEncoding)("cl100k_base");
        this.defaultTokenizer = {
            encode: (text)=>{
                return new Uint32Array(encoding.encode(text));
            },
            decode: (tokens)=>{
                const numberArray = Array.from(tokens);
                const text = encoding.decode(numberArray);
                const uint8Array = new TextEncoder().encode(text);
                return new TextDecoder().decode(uint8Array);
            }
        };
    }
    tokenizer(encoding) {
        if (encoding && encoding !== _types.Tokenizers.CL100K_BASE) {
            throw new Error(`Tokenizer encoding ${encoding} not yet supported`);
        }
        return this.defaultTokenizer;
    }
}
const tokenizers = new TokenizerSingleton();
